### PR TITLE
Step-7: Add filtering based on task state.

### PR DESCRIPTION
**Pull request by sta-wtag**
### [Step-7](https://trello.com/c/AdYcgZFe/6-step-7-add-filtering-based-on-task-state)

Add Task filtering to app.

### **Changes**
- [x] By default, show all the cards and make all as default filter.
- [x] Upon click on Incomplete , show only the incomplete tasks and same for Complete and All.
- [x] Make filtering play nice with pagination.

### CODE REVIEW

File: store/todos.js [hunk 1]
```diff
@@ -124,4 +139,33 @@ export const mutations = {
     state.perPage = 9;
     state.page = 1;
   },
+  setListPerPage(state, val) {
+    state.taskListPerPage = state.taskList;
+  },
+  filterTaskList(state, val) {
+    state.filterOptions.map((option) => (option.status = false));
```
Comment (hhu-wtag) [inline store/todos.js]: What are we mapping this into?

File: store/todos.js [hunk 2]
```diff
@@ -124,4 +139,33 @@ export const mutations = {
...
+    const option = state.filterOptions.find((option) => option.id === val.id);
+
+    option.status = true;
+
+    if (val.title === 'All') {
```
Comment (hhu-wtag) [inline store/todos.js]: Please use a CONST here. 
```suggestion
    import { ALL, INCOMPLETE, COMPLETE } from "@/utils/constants.js"

    if (val.title === ALL) {
```

File: store/todos.js [hunk 3]
```diff
@@ -124,4 +139,33 @@ export const mutations = {
...
+      state.taskListPerPage = state.taskList;
+
+      return;
+    }
+
+    if (val.title === 'Incomplete') {
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

File: store/todos.js [hunk 4]
```diff
@@ -124,4 +139,33 @@ export const mutations = {
...
+      state.taskListPerPage = state.taskList.filter(
+        (task) => task.done === false
+      );
+
+      return;
+    }
+
+    if (val.title === 'Complete') {
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: store/todos.js (hunk 1 above)
Comment (sta-wtag) [reply store/todos.js]: mapping option.status to false in every element

File: store/todos.js (hunk 2 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js (hunk 4 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js (hunk 3 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js (hunk 1 above)
Comment (hhu-wtag) [reply store/todos.js]: We can use forEach if we want change an array inline. Map returns a new mapped array which we aren't using here.

Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js [hunk 5]
```diff
@@ -1,5 +1,14 @@
...
-import { Limit } from '@/constants.js';
+import { forEach } from 'lodash';
+import {
+  Limit,
+  // eslint-disable-next-line camelcase
+  All_Task,
+  // eslint-disable-next-line camelcase
+  Incomplete_Task,
+  // eslint-disable-next-line camelcase
+  Complete_Task,
```
Comment (hhu-wtag) [inline store/todos.js]: Constant files should be uppercased. This will also resolve eslint errors

File: store/todos.js [hunk 6]
```diff
@@ -124,4 +148,36 @@ export const mutations = {
...
+    state.taskListPerPage = state.taskList;
+  },
+  filterTaskList(state, val) {
+    state.filterOptions.forEach((element) => (element.status = false));
+    const option = state.filterOptions.find((option) => option.id === val.id);
+
+    option.status = true;
+
+    // eslint-disable-next-line camelcase
+    if (val.title === All_Task) {
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

File: store/todos.js [hunk 7]
```diff
@@ -124,4 +148,36 @@ export const mutations = {
...
+      state.taskListPerPage = state.taskList;
+
+      return;
+    }
+
+    // eslint-disable-next-line camelcase
+    if (val.title === Incomplete_Task) {
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

File: store/todos.js [hunk 8]
```diff
@@ -124,4 +148,36 @@ export const mutations = {
...
+      state.taskListPerPage = state.taskList.filter(
+        (task) => task.done === false
+      );
+
+      return;
+    }
+
+    // eslint-disable-next-line camelcase
+    if (val.title === Complete_Task) {
```
Comment (hhu-wtag) [inline store/todos.js]: ditto

**Review by hhu-wtag — APPROVED** [review]

File: store/todos.js [hunk 9]
```diff
@@ -1,26 +1,44 @@
...
+import { forEach } from 'lodash';
+import { COMPLETE_TASK } from '../constants';
+import {
+  LIMIT,
+  // eslint-disable-next-line camelcase
+  All_Task,
+  // eslint-disable-next-line camelcase
+  Incomplete_Task,
+  // eslint-disable-next-line camelcase
+  Complete_Task,
```
Comment (bia-wtag) [inline store/todos.js]: please remove eslint rule disable for these lines and use UPPERCASE letters.

File: store/todos.js [hunk 10]
```diff
@@ -1,26 +1,44 @@
...
   page: 1,
   totalPage: 1,
   completeRequest: false,
   taskList: [],
-  filteredList: [],
+  taskListPerPage: [],
+  filterOptions: [
+    { id: uuid(), title: 'All', status: true },
+    { id: uuid(), title: 'Incomplete', status: false },
+    { id: uuid(), title: 'Complete', status: false },
```
Comment (bia-wtag) [inline store/todos.js:23]: should be translated.

File: store/todos.js [hunk 11]
```diff
@@ -124,4 +149,36 @@ export const mutations = {
...
+    if (val.title === INCOMPLETE_TASK) {
+      state.taskListPerPage = state.taskList.filter(
+        (task) => task.done === false
+      );
+
+      return;
+    }
+
+    // eslint-disable-next-line camelcase
+    if (val.title === COMPLETE_TASK) {
```
Comment (bia-wtag) [inline store/todos.js:174]: why these rules are ignored?

**Comment by bia-wtag** [comment]
@sta-wtag apu, couldn't run locally. Fails with the following errors. Also please consider using `yarn.lock` or `package-lock.json`, but not both as it'll create a problem with versions. Please have a look, thanks.
Screenshot 2023-02-08 at 9 32 16 AM
Screenshot 2023-02-08 at 9 31 54 AM
Screenshot 2023-02-08 at 9 31 41 AM

File: store/todos.js (hunk 11 above)
Comment (sta-wtag) [reply store/todos.js:174]: I forgot to remove the commented section after fixing the constant name to all uppercase

File: store/todos.js (hunk 9 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js (hunk 10 above)
Comment (sta-wtag) [reply store/todos.js:23]: Resolved

File: store/todos.js (hunk 8 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js (hunk 7 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js (hunk 6 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

File: store/todos.js (hunk 5 above)
Comment (sta-wtag) [reply store/todos.js]: Resolved

**Review by sai-wtag — APPROVED** [review]
Locally Tested. LGTM. Thanks

File: store/todos.js [hunk 12]
```diff
@@ -1,5 +1,12 @@
 import { uuid } from 'uuidv4';
```
Comment (bia-wtag) [inline store/todos.js:1]: nitpicking: this warning was shown in terminal, which might be related to this import.

```bash
(node:10626) DeprecationWarning: uuidv4() is deprecated. Use v4() from the uuid module instead.   09:04:24
(Use `node --trace-deprecation ...` to show where the warning was created)
```

File: store/todos.js [hunk 13]
```diff
@@ -124,4 +148,33 @@ export const mutations = {
     state.perPage = PER_PAGE;
     state.page = 1;
   },
+  setListPerPage(state, val) {
+    state.taskListPerPage = state.taskList;
+  },
+  filterTaskList(state, val) {
```
Comment (bia-wtag) [inline store/todos.js:154]: nitpicking: we might have better names than val 👀 i.e. `filter`

**Review by bia-wtag — APPROVED** [review]
Tested locally. LGTM 🚀

File: store/todos.js (hunk 12 above)
Comment (sta-wtag) [reply store/todos.js:1]: Okay. Thank you

File: store/todos.js (hunk 13 above)
Comment (sta-wtag) [reply store/todos.js:154]: Okay. Thank you

**Pull request merged**